src/transformStateWithClones.js

```diff
@@ -8,6 +8,44 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  const newState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        removeProperties(newState, action.keysToRemove);
+        break;
+
+      case 'clear':
+        clearProperties(newState);
+        break;
+    }
+
+    stateHistory.push({ ...newState });
+  }
+
+  return stateHistory;
+}
+
+function addProperties(state, extraData) {
+  Object.assign(state, extraData);
+}
+
+function removeProperties(state, keysToRemove) {
+  for (const key of keysToRemove) {
+    delete state[key];
+  }
+}
+
+function clearProperties(state) {
+  for (const key in state) {
+    delete state[key];
+  }
 }
```

Comment:
[CODE STYLE] - Instead of having separate functions for each action, you can move the logic for `addProperties`, `removeProperties`, and `clearProperties` directly into the respective `case` blocks of the `switch` statement. This will help you adhere to the principle of not repeating yourself as the action is performed once and only within the context of the `transformStateWithClones` function.